theano/tensor/opt.py

```diff
@@ -6223,6 +6223,33 @@ def add_calculate(num, denum, aslist=False, out_type=None):
 register_canonicalize(local_add_canonizer, name='local_add_canonizer')
 
 
+@register_canonicalize
```

Comment:
Where to register it is complicated. We need to register after the local_greedy_distributor optimization as otherwise, having local_mul_exp before/at the same time could remove some simplification done by local_greedy_distributor. As local_greedy_distributor is in canonicalize, this new opt can't be there.

This new opt should be done before local_pow_specialize that is is specialize, so it can't be in specialize.

Ideally, we don't want to add an extra pass on the graph between those 2 pass. We shouldn't add it in stabilize as this isn't stabilization.

So in conclusion, remove that line and use that diff to register it. This add it in an existing pass that isn't related, so they won't colide.

```
diff --git a/theano/tensor/opt.py b/theano/tensor/opt.py
index a00af70..5f57e21 100644
--- a/theano/tensor/opt.py
+++ b/theano/tensor/opt.py
@@ -5737,7 +5739,6 @@ def local_pow_canonicalize(node):
             return [broadcast_like(node.inputs[0], node.outputs[0], node.fgraph)]
     else:
         return False
-register_canonicalize(local_pow_canonicalize)
 
 
 @register_specialize
@@ -6193,7 +6194,11 @@ def local_log_sum_exp(node):
 
 
 compile.optdb.register('local_log_sum_exp',
-                       in2out(local_log_sum_exp, ignore_newtrees=True),
+                       in2out([local_log_sum_exp, local_mul_exp],
+                              ignore_newtrees=True),
+                       # between stabilize at 1.5 and specialize at 2
+                       # local_mul_exp should be after local_greedy_distributor
+                       # and before local_pow_specialize
                        1.6, 'fast_run')
 
```